src/psyclone/psyir/symbols/symboltable.py

```diff
@@ -180,40 +221,124 @@ def _normalize(key):
+
+        available_name = self.next_available_name(root_name, shadowing)
+        symbol = symbol_type(available_name, **symbol_init_args)
+        self.add(symbol, tag)
+        return symbol
+
+    def symbol_from_tag(self, tag, root_name=None, **new_symbol_args):
+        ''' Lookup a tag, if it doesn't exist create a new symbol with the
+        given tag. By default it creates a generic Symbol with the tag as the
+        root of the symbol name. Optionally, a different root_name or any of
+        the arguments available in the new_symbol() method can be given to
+        refine the name and the type of the created Symbol.
+
+        :param str tag: tag identifier.
+        :param str root_name: optional name of the new symbols if it needs \
+            to be created. Otherwise it is ignored.
+        :param new_symbol_args: arguments to create a new symbol.
+        :type new_symbol_args: unwrapped Dict[str, object]
+
+        :returns: symbol associated with the given tag.
+        :rtype: :py:class:`psyclone.psyir.symbols.Symbol`
+        '''
+        try:
+            symbol = self.lookup_with_tag(tag)
+            # Check that the symbol found matches the requested description
+            if 'symbol_type' in new_symbol_args:
+                symbol_type = new_symbol_args['symbol_type']
+                if not isinstance(symbol, new_symbol_args['symbol_type']):
+                    raise SymbolError(
+                        "Expected symbol with tag '{0}' to be of type '{1}' "
+                        "but found type '{2}'.".format(
+                        tag, symbol_type.__name__, type(symbol).__name__))
+            # TODO: What happens if the symbol is found and some unmatching
+            # arguments were given? If it needs to fail, we can implement the
+            # previous check polymorphically inside each symbol:
+            # symbol.match(arguments)
+            # which can also be useful for symbol comparisons in general.
+            return symbol
+        except KeyError:
+            if not root_name:
+                root_name = tag
+            return self.new_symbol(root_name, tag, **new_symbol_args)
+
+    def next_available_name(self, root_name=None, shadowing=False):
+        '''Return a name that is not in the symbol table and therefore can
+        be used to declare a new symbol.
+        If the `root_name` argument is not supplied or if it is
         an empty string then the name is generated internally,
         otherwise the `root_name` is used. If required, an additional
         integer is appended to avoid clashes.
+        If the shadowing argument is True (is False by default), the names
+        in parent symbol tables will not be considered.
 
         :param root_name: optional name to use when creating a new \
             symbol name. This will be appended with an integer if the name \
             clashes with an existing symbol name.
         :type root_name: str or NoneType
```

Comment:
I think we could just have `:param str root_name:` and drop this separate type description.

Reply:
Done